### PR TITLE
feat(ssrf): move schemes to shared include data

### PR DESCRIPTION
**Pull request by fzipi**
Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

Move the embedded list of schemes so it can be used by other RFI rules. Also we keep the list centralized.

### CODE REVIEW

File: util/regexp-assemble/data/934120.data [hunk 1]
```diff
@@ -14,8 +14,13 @@
 ##! - `##!$`: the suffix comment
 ##! Please refer to util/regexp-assemble/README.md for a full explanation.
 
-##! Plenty of schemes/protocols to catch here (from https://en.wikipedia.org/wiki/List_of_URI_schemes):
-##!^(?i)(?:acap|bitcoin|blob|cap|cvs|svn|svn\+ssh|turn|udp|vnc|xmpp|webcal|xri|dav|dns|feed|finger|h323|icap|ipp|ipps|jabber|mms|mumble|maven|nfs|rmi|rsync|rtmp|rtm|rtmfp|http|https|ftp|ftps|git|gopher|irc|fd|ldap|ldapi|ldaps|expect|netdoc|jar|ogg|tftp|dict|ssh2|ssh|imap|pop2|pop3|mailto|nntp|news|snews|snmp|smtp|sftp|telnet|s3|phar|sip|sips|smb)://
+##! This regex start with a list of all the protocols that can be used to make a request
```
Comment (theseion) [inline util/regexp-assemble/data/934120.data:17]:
```suggestion
##! This regex starts with a list of all the schemes that can be used to make a request
```

File: util/regexp-assemble/data/934120.data [hunk 2]
```diff
@@ -14,8 +14,13 @@
...
+
+##!> assemble
+  ##!> include protocols
```
Comment (theseion) [inline util/regexp-assemble/data/934120.data:20]:
```suggestion
  ##!> include url-schemes
```

File: util/regexp-assemble/data/934120.data [hunk 3]
```diff
@@ -14,8 +14,13 @@
 ##! - `##!$`: the suffix comment
 ##! Please refer to util/regexp-assemble/README.md for a full explanation.
 
-##! Plenty of schemes/protocols to catch here (from https://en.wikipedia.org/wiki/List_of_URI_schemes):
-##!^(?i)(?:acap|bitcoin|blob|cap|cvs|svn|svn\+ssh|turn|udp|vnc|xmpp|webcal|xri|dav|dns|feed|finger|h323|icap|ipp|ipps|jabber|mms|mumble|maven|nfs|rmi|rsync|rtmp|rtm|rtmfp|http|https|ftp|ftps|git|gopher|irc|fd|ldap|ldapi|ldaps|expect|netdoc|jar|ogg|tftp|dict|ssh2|ssh|imap|pop2|pop3|mailto|nntp|news|snews|snmp|smtp|sftp|telnet|s3|phar|sip|sips|smb)://
```
Comment (theseion) [inline util/regexp-assemble/data/934120.data:18]: Missing case flag: `##!+ i`

File: util/regexp-assemble/data/934120.data [hunk 4]
```diff
@@ -56,7 +61,7 @@
 [a-zA-Z][\w\-\.]{1,255}:\d{1,5}\+&@(?:\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}|[a-zA-Z][\w\-\.]{1,255}):\d{1,5}\/?\#\+@(?:\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}|[a-zA-Z][\w\-\.]{1,255}):\d{1,5}\/?
```
Comment (theseion) [inline util/regexp-assemble/data/934120.data:61]: Split this into multiple lines, using more concatenation comments.

**Review by theseion — CHANGES_REQUESTED** [review]
Please rename `protocols.data` to `url-schemes.data`. "Scheme" is the term used by the RFC.

File: util/regexp-assemble/data/include/protocols.data [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+##! Five special comments are at your disposal to influence the assembled expression:
+##! - `##!+`: the flag comment
+##! - `##!^`: the prefix comment
+##! - `##!$`: the suffix comment
+##! - `##!>`: the preprocessor comment
+##! - `##!<`: the block preprocessor end comment
+##!
+##! Currently supported preprocessors:
+##! - cmdline [windows | unix] (file scope)
+##! Please refer to util/regexp-assemble/README.md for a full explanation.
```
Comment (theseion) [inline util/regexp-assemble/data/include/protocols.data:22]: This is the new header that is currently waiting to be merged:
```
##! This is a data file used to generate a regular expression for a CRS rule.
##! The generation of the regular expression happens with the help of
##! util/regexp-assemble/regexp-assemble.py.
##! The ID of the rule in question is part of the file name of this data file.
##! Read more about the format of this data file and the use of the assembly
##! script in util/regexp-assemble/README.md.
##!
##! Lines starting with `##!` are comments and will be skipped,
##! empty lines will be ignored completely.
##! In addition, the quote character `'` at the beginning of a line will
##! cause the line to be interpreted literally by the cmdline preprocessor only.
##!
##! Seven special comments are at your disposal to influence the assembly of the
##! expression:
##! - `##!+`: the flag comment
##! - `##!^`: the prefix comment
##! - `##!$`: the suffix comment
##! - `##!>`: the preprocessor comment
##! - `##!<`: the preprocessor end comment
##! - `##!=>`: the concatenation output comment
##! - `##!=<`: the concatenation input comment
##!
##! Currently supported preprocessors:
##! - assemble
##! - cmdline
##! - template
##! - include
##!
##! Please refer to util/regexp-assemble/README.md for a full explanation.
```

File: util/regexp-assemble/data/934120.data (hunk 1 above)
Comment (fzipi) [reply util/regexp-assemble/data/934120.data:17]: Added.

File: util/regexp-assemble/data/934120.data (hunk 2 above)
Comment (fzipi) [reply util/regexp-assemble/data/934120.data:20]: Added.

File: util/regexp-assemble/data/934120.data (hunk 3 above)
Comment (fzipi) [reply util/regexp-assemble/data/934120.data:18]: Was in the other file, but moving it here for clarity.

File: util/regexp-assemble/data/include/protocols.data (hunk 5 above)
Comment (fzipi) [reply util/regexp-assemble/data/include/protocols.data:22]: Oh, ok. Will wait for that one then.

Comment (theseion) [reply util/regexp-assemble/data/include/protocols.data:22]: Actually, I meant that you could just put this into your file now and then there won't be any merge conflicts when I merge that PR.

Comment (fzipi) [reply util/regexp-assemble/data/include/protocols.data:22]: This will be moved to a unique file, so resolving.

**Comment by theseion** [comment]
I've replaced the complex IPv6 regular expressions with a single simple one. It doesn't make sense to replicate the specification, we only need to find strings that _look_ like IPv6 (the delimiting brackets `[`, `]` make this simple). In fact, replicating the spec would reduce performance and, potentially, open us up to more bypasses.

I've also expanded the "enclosed alphanumerics" matches to match all existing code points instead of matching only specific strings.

IMPORTANT: this PR relies on two PRs to fix bugs in regexp-assemble: #2624, #2623.

**Comment by theseion** [comment]
One thing I'm unsure about are evasions of the type `http://google.com:80+&@google.com:80#+@127.88.23.245:22/`. I don't know what the mechanism behind the evasion is. Currently, the `+` character is matched literally but I feel like it might actually be a URL encoded space character. Any ideas? @theMiddleBlue maybe?

**Comment by fzipi** [comment]
It is a space. See [this presentation](https://raw.githubusercontent.com/cujanovic/SSRF-Testing/master/us-17-Tsai-A-New-Era-Of-SSRF-Exploiting-URL-Parser-In-Trending-Programming-Languages.pdf)
image

**Comment by fzipi** [comment]
Ughhhhh.. We might want to include this one?
```
python3 -c 'import socket
host = """\\l\\o\\c\\a\\l\\h\\o\\s\\t"""
print(host)
print(socket.gethostbyname(host))'
\l\o\c\a\l\h\o\s\t
127.0.0.1
```

**Comment by theseion** [comment]
Requires a fix to Regexp::Assemble. Will updated ASAP.

**Comment by theseion** [comment]
@fzipi I have a working version (without having fixed the regexp-assemble bug yet).

**Review by theseion — APPROVED** [review]

**Comment by theseion** [comment]
I think there's still a duplicate in the tests but I wanted to get this to you as fast as possible.

**Comment by fzipi** [comment]
@theseion Awesome! But the bug is affecting the regex creation, right? In that case, are we doing to merge?

**Pull request merged**